First implementation

gulyasm opened the pull request:
_**Note:** This PR is not complete, I submitted it for early code review._
## What is in this PR?
- ID type with JSON serialization, deserialization, String method for human readable format. ID generation is statistically unique. The implementation comes from the appdash trace.
- SpanID for easy handling
- Event type
- Memory collector for testing purposes
- Trace RoundTripper implementation, that sets the ID for outgoing requests, reports the events.
- Trace method for instrumenting handlers. It sets the headers and record the events.
- Example changed to represent the new approach.
## What left to complete this PR?
- [x] Restructure code with reasonable structure. (header handling should go to `header.go`)
- [ ] Unit tests
- [x] Check naming. Some function names and the header names might be changed with better candidates.
- [x] Middleware for popular middleware systems.
- [x] Example on how to use it with middlewares (Negorni, Gorilla)
- [ ] Better documentation
## Other notes
- Part of the codebase (ID, SpanID) comes from the appdash trace library. (https://github.com/sourcegraph/appdash)

gulyasm committed:
feat(trace): Outgoing requests get SpanID.

New SpanID struct.
New ID struct to handle uint64 parsing, string transformation.
Trace Transpoort created with a SpanID assigns a SpanID to all outgoing
HTTP requests.

gulyasm committed:
feat(trace): Basic Memory collector added.

Memory collector added.
Collector stored in var for demonstration purposes.
RoundTripper records dummy event.

gulyasm committed:
feat(trace): Event type and JSON serialization added

Added ID custom JSON marshalling for messaging
Example changed to use an events endpoint for testing purposes. More
flexible and testable than the old timed version.
Added event recording for the middleware function and to the
RoundTripper.

gulyasm committed:
Negroni middleware added.

tmichel commented on example/basic.go:
```diff
@@ -0,0 +1,63 @@
+	"github.com/RisingStack/trace-go/trace"
+)
+
+var collector = trace.NewMemoryCollector()
```
I would have a "standard" collector in the package that is used by the `trace.XYZ` methods and functions we don't need to pass it around manually. And provide a method to override the default collector. Also provide a more finer grained approach - just like in this example.

I think the `log` and `http` packages are good examples for this. They have a sensible default as globally available and easy to use but you can have your own `Logger` instance.

tmichel commented on example/basic.go:
```diff
@@ -0,0 +1,63 @@
+	span := trace.NewSpanIDFromRequest(r)
+	log.Println("Id in 1st endpoint: " + span.String())
+	c := http.Client{
+		Transport: trace.Transport{
```
Can we make this API a little bit nicer? Something like:

``` go
c: = http.Client{Transport: trace.Transport(r)}
```

I think we can get the span id behind the scenes and the user only have to get the span id if she wants to use it.

gulyasm replied:
Done! :+1:

tmichel commented on trace/collector.go:
```diff
@@ -0,0 +1,49 @@
+	"sync"
+)
+
+type Collector interface {
```
:+1: for the simplicity.

tmichel commented on trace/collector.go:
```diff
@@ -0,0 +1,49 @@
+	return newList
+}
+
+func NewMemoryCollector() *MemoryCollector {
```
Only a nitpicking but I think that it is better to put the constructor function directly below the type definition:

``` go

type A struct {...}

func NewA(...) *A { ... }
```

I think it's easier to grok the given type.

gulyasm replied:
:+1:

tmichel commented on trace/collector.go:
```diff
@@ -0,0 +1,49 @@
+	return nil
+}
+
+func (c MemoryCollector) GetEvents() []Event {
```
I'm not 100% sure but I think if you are not using a pointer receiver then the `sync.Mutex` will be copied each time and on every method call you end up with a new mutex and locking will be non-effective. This is a problem because the `list` field is shared across all copies (as it is a slice).

gulyasm replied:
Yep, it was wrong. The `go vet` tool showed too.

tmichel commented on trace/collector.go:
```diff
@@ -0,0 +1,49 @@
+
+func NewMemoryCollector() *MemoryCollector {
+	c := MemoryCollector{}
+	c.list = []Event{}
```
You might want to initialize this slice with a biggish capacity so the copying can be minimized.

``` go
const NumberOfEventsBeforeFlush = 1000

...

c.list = make([]Event, 0, NumberOfEventsBeforeFlush)
```

gulyasm replied:
Nice catch! :+1:

tmichel commented on trace/event.go:
```diff
@@ -1,25 +1,33 @@
 
 const (
 	RequestReceived = iota
+	RequestCompleted
```
I would create a new type for these and implement a `String()` method on them and have it return the name of the constant for its value. It helps tremendously with debugging and if you use `int` you can still use `iota` for initialization. 

Also a custom `int` type would clarify the `Event` type and `NewEvent` function.

tmichel commented on trace/event.go:
```diff
@@ -1,25 +1,33 @@
 	return Event{
-		RequestId: uuid.NewV4().String(),
-		Type:      t,
 		CreatedAt: time.Now(),
```
I would say UTC is better than the local time.

tmichel commented on trace/http.go:
```diff
@@ -2,19 +2,71 @@ package trace
-	return c.Do(req)
-}
+const (
+	HeaderSpanID  = "Trace-Span"
```
Why did you leave the X prefix off? It had it in the initial version.

gulyasm replied:
`X-` prefixes are now discouraged: http://stackoverflow.com/questions/3561381/custom-http-headers-naming-conventions

tmichel commented on trace/id.go:
```diff
@@ -0,0 +1,142 @@
+}
+
+func (i *ID) MarshalJSON() ([]byte, error) {
+	return []byte("\"" + i.String() + "\""), nil
```
I think it would help the readability if you used `fmt.Sprintf` instead of string concatencation.

``` go
return []byte(fmt.Sprintf("%q", id.String())
```

gulyasm replied:
:+1:

tmichel commented:
Awesome PR. I had some minor comments.

The public API of the package might need some further fine tuning but I think the general direction is good.

Implementing a standard `http.Handler` would mean it can be integrated with any http routing/middleware lib that uses the standard `http.Server`. The user could just wrap the base handler that is passed to `http.ListendAndServe` with a trace middleware.

Another thing is that I don't see a way to get an already initialized span id for a given request. Did I miss it or is it missing?

gulyasm committed:
feat(trace): Added base handler for general instrumentation

gulyasm committed:
feat(trace): Added default values for easier integration.

gulyasm committed:
feat(trace): Code repetation removed

gulyasm committed:
fix(build): Fixed make test error.

gulyasm committed:
test(trace): Added missing unit tests for IDs.

gulyasm committed:
refactor(trace): Refactored for clarity.

gulyasm committed:
feat(trace): Added negroni integration

gulyasm committed:
docs(trace): Added documentation for ID, SpanID, and Trace.

gulyasm committed:
fix(trace): Events are now created with UTC time.

gulyasm commented:
@tmichel The design is final by me. Could you review once again please? I will add more and more tests, but I plan to modify design only if there is feedback to do so. 

I think I found a pretty clean implementation and it's also easy to add new integration. Right now you can see the integrations in the examples. 

The code is organized as the following:
http.go - Trace
client.go - Custom Transport
id.go - The SpanID and ID
header.go - The parsing and request setting
event.go - The trace event
collector.go - The collectors, right now, only the MemoryCollector.

As far as I can see I added all your suggestions (except the EventType type, I would add that later on this week), they were awesome-great! Also, I ended up with the current design when I looked into your http.Handler suggestion. Big :+1: !

I keep adding tests and plan to add Integration doc later this week. 

cc @gergelyke

gulyasm committed:
refactor(trace): Added refactors from code review comments.

tmichel commented on trace/collector.go:
```diff
@@ -0,0 +1,51 @@
+}
+
+func (c *MemoryCollector) GetEvents() []Event {
+	c.lock.Lock()
```
You could use the `sync.RWMutex` and allow reads from multiple threads.

tmichel commented on trace/header.go:
```diff
@@ -1,34 +1,60 @@
 
-func GetId(r *http.Request) string {
-	return r.Header.Get(HeaderId)
+// Generates a new SpanID from a request. If there is no span information in the request, a root SpanID will be generated.
```
The function/method documentation usually starts with the name of the function/method. 

> Notice this comment is a complete sentence that begins with the name of the element it describes. This important convention allows us to generate documentation in a variety of formats, from plain text to HTML to UNIX man pages, and makes it read better when tools truncate it for brevity, such as when they extract the first line or sentence.

Source: http://blog.golang.org/godoc-documenting-go-code

tmichel commented on trace/header.go:
```diff
@@ -1,34 +1,60 @@
+	traceIDStr := req.Header.Get(HeaderTraceID)
+	spanID, err := ParseID(spanIDStr)
+	if err != nil {
+		log.Println("Failed to parse spanID string value, using 0.")
```
I think logging in a library is very annoying. Especially if you don't provide any control over it. What if I want to use a different logger. What I've seen recently is to create an interface that matches the stdlib's logger and make it settable. The default could be an instance of the stdlib's logger. [Something like this](https://github.com/Shopify/sarama/blob/7cdfc5b3e054fe72ad79af2a2984bbffcfcf3c64/sarama.go#L23-L30).

Also if you ignore the error please be explicit about the fallback `spanID`. This could hide some nasty bugs if the the return value of `ParseID` changes in the error case. Also this introduces some tight coupling.

gulyasm replied:
I documented this behaviour. The good solution should be to differentiate between missing header and failed parsing. I will change this behaviour later.

tmichel commented on trace/http.go:
```diff
@@ -2,19 +2,46 @@ package trace
+}
+
+// Handler wraps an existing http.Handler with Trace functionality.
+func (t *Trace) Handler(h http.Handler) http.Handler {
```
I would change the signature here to Handler be just a simple function. I think generating an `http.Handler` that uses a tracer is not the responsibility of the `Trace` object.

I would do something like:

``` go
func Handler(t *Trace, h http.Handler) http.Handler) {
    ....
}
```

And then use it like:

``` go
http.ListenAndServe(":8080", trace.Handler(t, mux))
```

tmichel commented on trace/http.go:
```diff
@@ -2,19 +2,46 @@ package trace
+	collector Collector
+}
+
+func (t *Trace) process(rw http.ResponseWriter, r *http.Request, f http.HandlerFunc) {
```
It would be cleaner and more general to take an `http.Handler` instead of an `http.HandlerFunc`.

tmichel commented on trace/http.go:
```diff
@@ -2,19 +2,46 @@ package trace
 
-func Instrument(fn http.HandlerFunc) http.HandlerFunc {
+// HandlerFunc wraps an existing http.HandlerFunc with Trace functionality.
+func (t *Trace) HandlerFunc(f http.HandlerFunc) http.HandlerFunc {
```
I think this is unnecessary and the `Handler` function should be enough. You would rarely wrap a function.

tmichel commented on trace/http.go:
```diff
@@ -2,19 +2,46 @@ package trace
 }
+
+// HandlerFuncNext middleware function can be used for Negroni integration.
+func (t *Trace) HandlerFuncNext(rw http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
```
The naming is a bit strange. If this is really for the negroni integration then name it so it is obvious. 
And I would go for the function taking a `*Trace` or `Trace` instead of a method on `Trace`.

So

``` go
trace.Negroni(...)
```

is more straigtforward than

``` go
t.HandlerFuncNext(...)
```

Although implementing the `http.Handler` should be more than enough to integrate with almost every web framework/middleware lib out there. 

You could just modify the negroni hello world example:

``` go
package main

import (
  "github.com/codegangsta/negroni"
  "net/http"
  "fmt"
)

func main() {
  mux := http.NewServeMux()
  mux.HandleFunc("/", func(w http.ResponseWriter, req *http.Request) {
    fmt.Fprintf(w, "Welcome to the home page!")
  })

  t := trace.NewTrace()
  n := negroni.Classic()
  n.UseHandler(trace.Handler(t, mux))
  n.Run(":3000")
}
```

This reduces the API surface which imo is always good. You could have a section in the readme or in `doc.go` that showcases how to integrate with a few popular libraries.

tmichel commented on trace/id.go:
```diff
@@ -0,0 +1,134 @@
+}
+
+// Empty return true if this SpanID is empty, elements are 0s.
+func (s *SpanID) Empty() bool {
```
Maybe `IsEmpty()`? For me `Empty()` means that it returns an empty instance.

tmichel commented on trace/id_test.go:
```diff
@@ -0,0 +1,156 @@
+	}
+}
+
+func TestString(t *testing.T) {
```
This test method does not say anything about what it tests.

What I usually do is include the type in the test function name so it is obvious what is the subject of the test.

tmichel commented on trace/id_test.go:
```diff
@@ -0,0 +1,156 @@
+
+func TestRootIsRoot(t *testing.T) {
+	i := NewRootSpanID()
+	log.Println(i.Empty())
```
Please don't log in tests. If you really have to log something then use `t.Log` and `t.Logf` which would only show up if you run `go test -v`.

gulyasm replied:
This is somehow an outdated diff. I removed debug logs.

tmichel commented on try:
```diff
@@ -0,0 +1,2 @@
+#!/bin/bash
```
I'm sure this file was committed by accident. :smile:

tmichel commented:
I had some minor comments but the API is sound.

The only question for me is integration with other libs. I would keep the base package really simple and the API as small as possible. So not include any special integration for Negroni or anything else. `http.Handler` is all you need in most cases. You could always add more packages to this repo for integrating with libs/framworks but that could mess up vendoring. I've run into this recently and it was not straightforward to clean it up. So maybe just have different repos if the integration is so complex. A good readme/wiki with lots of integrations are a better alternative for me.

gulyasm committed:
tests(trace): Added unit tests for http and header.

gulyasm committed:
refactor(trace): Code review refactor.

Empty function renamed to IsEmpty for clarity.
TestString method name renamed to TestIDString for better debug message.
String concanation changed to fmt.Sprintf

gulyasm committed:
docs(trace): Added documentation for header.go

gulyasm committed:
refactor(trace): Removed accidentally committed try file.

gulyasm committed:
feat(trace): Removed Negroni. Added RWLock.

Removed Negroni HandlerFuncNext, it might be added as a seperate
package.
MemoryCollector Lock changed to RWLock.

gulyasm committed:
feat(trace): Removed negroni example.

gulyasm commented:
I will close this PR if no comment comes to this until Tuesday. I will add integration docs and unit tests in another, seperate PR as this one is already very long. The API is final so it can be merged to the master. 
:metal: :sunglasses: 

@tmichel I left the signature of the Handler func as it is, because some middleware library like Alice (https://github.com/justinas/alice) supports the `func(http.Handler) http.Handler` signature.
Removed the Negroni func, it can be added as a seperate library, but left the HandlerFunc because the syntax to solve wrapping a HandlerFunc is really wierd with just a Handler.

tmichel commented:
Everything LGTM. Should I merge?

gulyasm commented:
Yes please, and thanks! :clap:
